stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,143 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+
+struct Boxed[T: AnyType, address_space: AddressSpace = AddressSpace.GENERIC]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist. Consider Boxed[T] over UnsafePointer[T] where possible.
+
+    Parameters:
+        T: The type to be stored in the Boxed[].
+        address_space: The address space that the allocation behind this box resides in.
+    """
+
+    var _inner: UnsafePointer[T, address_space]
+
+    fn __init__[
+        T: Movable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
+        """Construct a new Boxed[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, AddressSpace.GENERIC].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_value: T):
+        """Construct a new Boxed[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, address_space].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](
+        inout self: Boxed[T, AddressSpace.GENERIC],
+        *,
+        copy_box: Boxed[T, AddressSpace.GENERIC],
+    ):
+        """Construct a new Boxed[] by explicitly copying the value from another Boxed[].
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_box: The Boxed[] to copy.
+        """
+        self.__init__(copy_value=copy_box[])
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Move this Boxed[].
+
+        Args:
+            existing: The value to move.
+        """
+        self._inner = existing._inner
+        existing._inner = UnsafePointer[T, address_space]()
+
+    fn __getitem__(
+        ref [_, address_space._value.value]self
+    ) -> ref [__lifetime_of(self._inner), address_space._value.value] T:
```

Comment:
@lattner will the lifetime system "do the right thing" here?  Is `__lifetime_of(self._inner)` the same as `__lifetime_of(self)` here?  How should we think about tying the return type of public APIs with a lifetime associated with a field (and a private field at that!) from a style perspective?